File: web/src/lib/hierarchy.ts
/**
 * API client and types for Project Hierarchy (Epic > Story > Task)
 */

import { apiFetch } from "./api";

export type ItemType = "story" | "task" | "feature" | "bug";

export interface HierarchyItem {
  id: number;
  item_node_id: string;
  title: string | null;
  item_type: ItemType | null;
  status: string | null;
  epic_name: string | null;
  parent_item_id: number | null;
  labels: string[] | null;
  children: HierarchyItem[];
}

export interface HierarchyEpic {
  epic_option_id: string | null;
  epic_name: string | null;
  items: HierarchyItem[];
}

export interface HierarchyResponse {
  epics: HierarchyEpic[];
  orphans: HierarchyItem[];
}

/**
 * Get the complete hierarchy for the current project
 */
export async function getProjectHierarchy(): Promise<HierarchyResponse> {
  return apiFetch<HierarchyResponse>(`/api/projects/current/hierarchy`);
}

/**
 * Helper to flatten hierarchy into a list
 */
export function flattenHierarchy(hierarchy: HierarchyResponse): HierarchyItem[] {
  const items: HierarchyItem[] = [];

  function collectItems(item: HierarchyItem) {
    items.push(item);
    item.children.forEach(collectItems);
  }

  hierarchy.epics.forEach((epic) => {
    epic.items.forEach(collectItems);
  });

  hierarchy.orphans.forEach(collectItems);

  return items;
}

/**
 * Helper to count items by type
 */
export function countByType(hierarchy: HierarchyResponse): Record<string, number> {
  const items = flattenHierarchy(hierarchy);
  const counts: Record<string, number> = {};

  items.forEach((item) => {
    const type = item.item_type || "undefined";
    counts[type] = (counts[type] || 0) + 1;
  });

  return counts;
}

/**
 * Helper to get all stories
 */
export function getStories(hierarchy: HierarchyResponse): HierarchyItem[] {
  return flattenHierarchy(hierarchy).filter((item) => item.item_type === "story");
}

/**
 * Helper to get all tasks
 */
export function getTasks(hierarchy: HierarchyResponse): HierarchyItem[] {
  return flattenHierarchy(hierarchy).filter(
    (item) => item.item_type === "task" || item.item_type === "feature" || item.item_type === "bug"
  );
}

File: api/app/schemas/hierarchy.py
from pydantic import BaseModel, Field


class HierarchyItemResponse(BaseModel):
    """Schema for a single item in the hierarchy tree."""
    id: int
    item_node_id: str
    title: str | None
    item_type: str | None = Field(None, description="story, task, feature, bug")
    status: str | None
    epic_name: str | None
    parent_item_id: int | None
    labels: list[str] | None
    children: list["HierarchyItemResponse"] = Field(default_factory=list)

    class Config:
        from_attributes = True


class HierarchyEpicResponse(BaseModel):
    """Schema for an epic group in the hierarchy."""
    epic_option_id: str | None
    epic_name: str | None
    items: list[HierarchyItemResponse]


class HierarchyResponse(BaseModel):
    """Schema for complete hierarchy response."""
    epics: list[HierarchyEpicResponse]
    orphans: list[HierarchyItemResponse] = Field(
        default_factory=list,
        description="Items sem épico definido"
    )


# Needed for self-referential model
HierarchyItemResponse.model_rebuild()
